pyro/contrib/mue/variablelengthhmm.py

```diff
@@ -0,0 +1,87 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.hmm import _sequential_logmatmulexp
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.util import broadcast_shape
+
+
+class VariableLengthDiscreteHMM(TorchDistribution):
```

Comment:
This is a really cool distribution 👍 I'd consider renaming it to maybe `MissingDataDiscreteHMM` or `OneHotDiscreteHMM` since it is quite a bit more general than `VariableLength`. Once it settles down, we could even move it to `pyro.distributions`.